### PR TITLE
fix(core): fix policy synthesizer logic for precreated roles

### PR DESCRIPTION
**Pull request by GavinZZ**
### Issue # (if applicable)

Closes https://github.com/aws/aws-cdk/issues/31653

### Reason for this change

With Role.customizeRoles enabled, dynamodb.Table.addGlobalSecondaryIndex causes an error. This is a critical blocker for customers who require the use of customizeRoles.

#### Intended behaviour

When `customizeRoles` is used, the `iam-policy-report.txt` report will contain a list
of IAM roles and associated permissions that would have been created. This report is
generated so that it attempts to resolve any references and replace with a more user
friendly value.

The following are some examples of the value that will appear in the report:

```json
"Resource": {
      "Fn::GetAtt": [
           "SomeResource",
           "Arn"
       ]
}
```

The policy report will instead get:

```json
"(Path/To/SomeResource.Arn)"
```

#### Current issues
There are two main issues here:
1. Policy synthesizer (which is used for customizeRoles to generate report) is created with `App` scope. This caused the failure in the original issue `Resolution error: PolicySynthesizer at 'PolicySynthesizer' should be created in the scope of a Stack, but no Stack found.` because token resolution requires a Stack scope not an App scope.
2. The policy synthesizer was using `DefaultTokenResolver`. The default token resolution class does not generate the same format of output values for the policy report. i.e. A concatenated token value, i.e. `${Token[Token.X]}/index/*` would be converted to `(PhysicalId).Arn` instead of `"(Path/To/SomeResource.Arn)"`. 
3. Pseudo parameters like `AWS::NoValue` would be rendered as `Tokens` in the policy report which is not idea. Update it to make it output `NOVALUE`.

This PR addresses the above two issues.

### Description of how you validated changes

New and existing tests pass.

### Checklist
- [ ] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-iam/README.md [hunk 1]
```diff
@@ -238,6 +238,72 @@ iam.Role.customizeRoles(this, {
 For more information on configuring permissions see the [Security And Safety Dev
 Guide](https://github.com/aws/aws-cdk/wiki/Security-And-Safety-Dev-Guide)
 
+#### Policy report generation
+
+When `customizeRoles` is used, the `iam-policy-report.txt` report will contain a list
+of IAM roles and associated permissions that would have been created. This report is
+generated so that it attempts to resolve any references and replace with a more user
+friendly value.
```
Comment (Leo10Gama) [inline packages/aws-cdk-lib/aws-iam/README.md:246]: nit: phrasing feels awkward
```suggestion
When `customizeRoles` is used, the `iam-policy-report.txt` report will contain a list
of IAM roles and associated permissions that would have been created. This report is
generated in an attempt to resolve and replace any references with a more user-friendly
value.
```

**Review by Leo10Gama — CHANGES_REQUESTED** [review]
Overall changes look good! Just a few minor comments from me.

File: packages/@aws-cdk-testing/framework-integ/test/aws-dynamodb/test/integ.table-with-customized-role.js.snapshot/iam-policy-report.txt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      "dynamodb:ConditionCheckItem",
+      "dynamodb:DescribeTable"
+    ],
+    "Effect": "Allow",
+    "Resource": [
+      "(cdk-table-with-customized-role/Table/Resource.Arn)",
+      "(cdk-table-with-customized-role/Table/Resource.Arn)/index/*"
+    ]
+  }
+]<AWSServiceRoleForApplicationAutoScaling_DynamoDBTable> (cdk-table-with-customized-role2/Table/ScalingRole)
```
Comment (Leo10Gama) [inline packages/@aws-cdk-testing/framework-integ/test/aws-dynamodb/test/integ.table-with-customized-role.js.snapshot/iam-policy-report.txt:51]: I'm not entirely familiar with it, but is this expected behaviour for the policy report? It seems like here and on L13 and L63 there ought to be newlines separating the different roles.

Comment (GavinZZ) [reply packages/@aws-cdk-testing/framework-integ/test/aws-dynamodb/test/integ.table-with-customized-role.js.snapshot/iam-policy-report.txt]: Nice find! These are supposed two be two separate IAM roles and should have newlines between them. I think the original author made a mistake and did not split the lines correctly for multiple roles case. I've fixed it.

**Review by Leo10Gama — APPROVED** [review]
Changes LGTM! Will leave the `do-not-merge` label on for a second review from core to confirm changes are fine

**Comment by GavinZZ** [comment]
Discussed with Kendra offline and she's fine to merge this. Removing `do-not-merge`.

**Pull request merged**